doc/source/templates/04_finetuning_llms/finetune_llama.py

```diff
@@ -0,0 +1,438 @@
+        accelerator.print("Eval perplex", perplex)
+
+        eval_e_epoch = time.time()
+        accelerator.print("Eval time per epoch: ", eval_e_epoch - eval_s_epoch)
+        accelerator.print("avg fwd time: ", avg_fwd_time / (step + 1))
+        accelerator.print("avg bwd time: ", avg_bwd_time / (step + 1))
+        accelerator.print("avg opt step time: ", avg_opt_step_time / (step + 1))
+
+        session.report(
+            {
+                "epoch": epoch,
+                "train_loss": loss_sum.item() / train_ds_len,
+                "eval_loss": eloss.item() / train_ds_len,
+                "perplexity": perplex / valid_ds_len,
+                "number of iterations": step + 1,
+                "Train time per epoch": e_epoch - s_epoch,
+                "Eval time per epoch": eval_e_epoch - eval_s_epoch,
+                "avg fwd time": avg_fwd_time / (step + 1),
+                "avg bwd time": avg_bwd_time / (step + 1),
+            },
+        )
+        
+        if args.output_dir is not None:
+            accelerator.print(f"Saving bin version of model in {args.output_dir}")
+            accelerator.wait_for_everyone()
+            with accelerator.main_process_first():
+                if not os.path.exists(os.path.join(args.output_dir, str(epoch))):
+                    os.makedirs(os.path.join(args.output_dir, str(epoch)))
+            unwrapped_model = accelerator.unwrap_model(model)
+            unwrapped_model.save_pretrained(
+                os.path.join(args.output_dir, str(epoch)),
+                is_main_process=accelerator.is_main_process,
+                save_function=accelerator.save,
+                safe_serialization=False,
+                state_dict=accelerator.get_state_dict(model),
+            )
+
+def main():
+    global MODEL
+    global EVAL_BATCH_SIZE
+    
+    parser = argparse.ArgumentParser(description="Simple example of training script.")
+    parser.add_argument(
+        "--mx",
+        type=str,
+        default="bf16",
+        choices=["no", "fp16", "bf16", "fp8"],
+        help="Whether to use mixed precision. Choose"
+        "between fp16 and bf16 (bfloat16). Bf16 requires PyTorch >= 1.10."
+        "and an Nvidia Ampere GPU.",
+    )
+    parser.add_argument("--bs", type=int, default=64, help="Batch size to use.")
+    parser.add_argument("--grad_accum", type=int, default=1, help="Gradient accumulation steps.")
+    parser.add_argument("--train_path", type=str, help="Path to training jsonl file")
+    parser.add_argument("--test_path", type=str, help="Path to testing jsonl file")
+    parser.add_argument("--special_token_path", type=str, help="Path to token json file")    
+    parser.add_argument("--no-grad-ckpt", action="store_true", help="If passed, will not use gradient checkpointing.")
+    parser.add_argument("--output_dir", type=str, help="Path to output directory.")
+    parser.add_argument("--model_name", default="tiiuae/falcon-7b", type=str)
```

Comment:
nit: maybe rename to `model_name_or_path`, since we are loading llama-2 from local ckpt dir.